src/lib.rs

```diff
@@ -129,18 +152,77 @@ impl StdError for Error {
         match *self {
             AlignmentMismatch { .. } => "Alignment Mismatch",
             LengthMismatch { .. } => "Length Mismatch",
+            CapacityMismatch { .. } => "Capacity Mismatch",
         }
     }
 }
 
-fn check_constraints<U>(data: &[u8]) -> Result<usize, Error>
+trait Empty {
+    fn is_empty(&self) -> bool;
+}
+
+impl<T> Empty for Vec<T> {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+impl<T> Empty for [T] {
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+trait Len {
```

Comment:
Instead of these two traits you could work with `AsRef<[T]>` which is implemented by both types, and then call `len()` and `is_empty()` on the result of `as_ref()`. No need to add new traits :)

Reply:
Ah, Thanks I didn't know about this :-),  so the following signature would be fine?
```rust
fn check_length<T, U>(data: &T) -> Result<usize, Error>
where
    U: TypeName,
    T: AsRef<[u8]> + ?Sized,
```

Reply:
Yes